shared/js/participantStore.js

```diff
@@ -0,0 +1,104 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+var loop = loop || {};
+loop.store = loop.store || {};
+
+loop.store.ParticipantStore = (function() {
+  "use strict";
+
+  const PARTICIPANT_SCHEMA = {
+    name: "",
+    isHere: false,
+    pingTime: null
+  };
+
+  /**
+   * Participant store.
+   *
+   * @param {loop.Dispatcher} dispatcher  The dispatcher for dispatching actions
+   *                                      and registering to consume actions.
+   * @param {Object} options Options object:
+   * - {OTSdkDriver} sdkDriver  The SDK driver instance.
+   */
+  var ParticipantStore = loop.store.createStore({
+    initialize: function(options) {
+      if (!options.sdkDriver) {
+        throw new Error("Missing option sdkDriver");
+      }
+      this._sdkDriver = options.sdkDriver;
+    },
+
+    getInitialStoreState: function() {
+      return {
+        participants: new Map()
+      };
+    },
+
+    actions: [
+      "updatedParticipant",
+      "updatedPresence"
+    ],
+
+    /**
+     * Checks if the room is empty or has participants.
+     */
+    _hasParticipants: function() {
+      return this._storeState.participants.size > 0;
+    },
+
+    updatedParticipant: function(actionData) {
+      let participant = this._updateParticipantData(actionData.userId, {
+        name: actionData.name
+      });
+
+      this.setStoreState({
+        participants: this._storeState.participants.set(actionData.userId, participant)
+      });
+    },
+
+    updatedPresence: function(actionData) {
+      let participant = this._updateParticipantData(actionData.userId, {
+        isHere: actionData.isHere,
+        pingTime: actionData.pingTime
+      });
+
+      let updatedParticipants = this._storeState.participants.set(actionData.userId, participant);
+
+      this.setStoreState({
+        participants: updatedParticipants
```

Comment:
This being slightly different from the pattern in `updatedParticipant` is somewhat confusing. Might as well just move this logic into `_updateParticipantData`.